File: leetcode/cpp/generate_parentheses.cpp
class Solution {
	public:
	    vector<string> generateParenthesis(int n) {       
			vector<string> ret;
			if (n == 0) return ret;
			dfs(n, n, "", ret);
			return ret;
		}

		void dfs(int left, int right, string entry, vector<string> &ret) {
			if (left == 0 && right == 0) {
				ret.push_back(entry);
				return;
			}

			if (left < right)
				dfs(left, right-1, entry+")", ret);

			if (left != 0)
				dfs(left-1, right, entry+"(", ret);
		}
};

File: leetcode/java/generate_parentheses.java
public class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ret = new ArrayList<String>();
        dfs(ret, "", n, n);
        return ret;
    }

    private void dfs(List<String> ret, String entry, int left, int right) {
        if (right == 0) {
            ret.add(entry);
            return;
        }
        if (left > 0) {
            dfs(ret, entry + "(", left-1, right);
        }
        if (left < right) {
            dfs(ret, entry + ")", left, right-1);
        }
    }
}
